### PR TITLE
goftp: allow parsing of windows LIST statement

### PR DESCRIPTION
**Pull request by mountain-root**

### CODE REVIEW

File: file_system.go [hunk 1]
```diff
@@ -321,15 +321,27 @@ func (f *ftpFile) Sys() interface{} {
...
 	if strings.HasPrefix(entry, "total ") {
 		return nil, nil
 	}
 
+	if lsRegex.MatchString(entry) {
+		return parseLinuxLIST(entry, loc, skipSelfParent)
+	} else if winLsRegex.MatchString(entry) {
+		return parseWinLIST(entry, loc)
+	}
+	return nil, fmt.Errorf("unexpected LIST format %v", entry)
```
Comment (jwebb) [inline file_system.go:339]:
```suggestion
	return nil, ftpError{err: fmt.Errorf(`unexpected LIST format: %s`, entry)}
```

To match the other error results here.

File: file_system.go [hunk 2]
```diff
@@ -321,15 +321,27 @@ func (f *ftpFile) Sys() interface{} {
...
 
-// total 404456
-// drwxr-xr-x   8 goftp    20            272 Jul 28 05:03 git-ignored
 func parseLIST(entry string, loc *time.Location, skipSelfParent bool) (os.FileInfo, error) {
 	if strings.HasPrefix(entry, "total ") {
 		return nil, nil
 	}
 
+	if lsRegex.MatchString(entry) {
+		return parseLinuxLIST(entry, loc, skipSelfParent)
```
Comment (jwebb) [inline file_system.go:335]: Probably Unix, not Linux

File: file_system.go [hunk 3]
```diff
@@ -321,15 +321,27 @@ func (f *ftpFile) Sys() interface{} {
 	return f.raw
 }
 
-var lsRegex = regexp.MustCompile(`^\s*(\S)(\S{3})(\S{3})(\S{3})(?:\s+\S+){3}\s+(\d+)\s+(\w+\s+\d+)\s+([\d:]+)\s+(.+)$`)
+var (
+	lsRegex    = regexp.MustCompile(`^\s*(\S)(\S{3})(\S{3})(\S{3})(?:\s+\S+){3}\s+(\d+)\s+(\w+\s+\d+)\s+([\d:]+)\s+(.+)$`)
```
Comment (jwebb) [inline file_system.go:325]: Rename as well?

File: file_system.go [hunk 4]
```diff
@@ -394,6 +406,42 @@ func parseLIST(entry string, loc *time.Location, skipSelfParent bool) (os.FileIn
...
+		return nil, err
+	}
+
+	fi := matches[2]
+	mode := os.ModePerm
+	var size uint64
+	if fi == "<DIR>" {
+		mode |= os.ModeDir
+	} else {
+		mode |= os.ModeSymlink
```
Comment (jwebb) [inline file_system.go:429]: Not a symlink, just a regular file

File: file_system.go [hunk 5]
```diff
@@ -394,6 +406,42 @@ func parseLIST(entry string, loc *time.Location, skipSelfParent bool) (os.FileIn
...
+		size, err = strconv.ParseUint(fi, 10, 64)
+		if err != nil {
+			return nil, ftpError{err: fmt.Errorf(`failed parsing LIST entry's size: %s (%s)`, err, entry)}
+		}
+	}
+
+	return &ftpFile{
+		name:  filepath.Base(matches[3]),
```
Comment (jwebb) [inline file_system.go:437]: I don't think Windows would ever return a full path in a directory listing, so filepath.Base should be unnecessary (it's also probably incorrect in the existing code as well, as it's assuming the local OS path separator applies on the remote, but let's leave that alone here).

File: file_system.go (hunk 4 above)
Comment (mountain-root) [reply file_system.go:429]: regular file looks to be the default, so I can just remove this line correct/

Comment (jwebb) [reply file_system.go]: Yep

**Pull request merged**